pkg/controller/controller_instance.go

```diff
@@ -2021,3 +1830,216 @@ func shouldStartOrphanMitigation(statusCode int) bool {
 		statusCode == http.StatusRequestTimeout ||
 		is5XX
 }
+
+// serviceInstanceHasExistingBindings returns true if there are any existing
+// bindings associated with the given ServiceInstance.
+func (c *controller) serviceInstanceHasExistingBindings(instance *v1beta1.ServiceInstance) (bool, error) {
+	bindingLister := c.bindingLister.ServiceBindings(instance.Namespace)
+
+	selector := labels.NewSelector()
+	bindingList, err := bindingLister.List(selector)
+	if err != nil {
+		return false, err
+	}
+	for _, binding := range bindingList {
+		if instance.Name == binding.Spec.ServiceInstanceRef.Name {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// cloneServiceInstance clones the given ServiceInstance. This is so that the
+// reconciler may modify the cloned resource without accidentally mutating the
+// cache.
+func cloneServiceInstance(instance *v1beta1.ServiceInstance) (*v1beta1.ServiceInstance, error) {
+	clone, err := api.Scheme.DeepCopy(instance)
+	if err != nil {
+		return nil, err
+	}
+	return clone.(*v1beta1.ServiceInstance), nil
+}
+
+// updateServiceInstanceGracefulDeletionSuccess handles the logging and
+// updating of a ServiceInstance that has successfully finished graceful
+// deletion.
+func (c *controller) updateServiceInstanceGracefulDeletionSuccess(instance *v1beta1.ServiceInstance) error {
+	pcb := pretty.NewContextBuilder(pretty.ServiceInstance, instance.Namespace, instance.Name)
+
+	instance, err := cloneServiceInstance(instance)
+	if err != nil {
+		return fmt.Errorf(pcb.Messagef("error cloning: %v", err))
+	}
+
+	clearServiceInstanceCurrentOperation(instance)
+
+	finalizers := sets.NewString(instance.Finalizers...)
+	finalizers.Delete(v1beta1.FinalizerServiceCatalog)
+	instance.Finalizers = finalizers.List()
+
+	if _, err := c.updateServiceInstanceStatus(instance); err != nil {
+		return err
+	}
+
+	glog.Info(pcb.Message("Cleared finalizer"))
+
+	return nil
+}
+
+// updateServiceInstanceReconciliationError handles the logging and updating of
+// a ServiceInstance that hit a retryable error during reconciliation.
+func (c *controller) updateServiceInstanceReconciliationError(instance *v1beta1.ServiceInstance, reason, message string) error {
+	pcb := pretty.NewContextBuilder(pretty.ServiceInstance, instance.Namespace, instance.Name)
+
+	instance, err := cloneServiceInstance(instance)
+	if err != nil {
+		return fmt.Errorf(pcb.Messagef("error cloning: %v", err))
+	}
+
+	setServiceInstanceCondition(
+		instance,
+		v1beta1.ServiceInstanceConditionReady,
+		v1beta1.ConditionFalse,
+		reason,
+		message,
+	)
+	if _, err := c.updateServiceInstanceStatus(instance); err != nil {
+		return err
+	}
+
+	glog.Info(pcb.Message(message))
```

Comment:
Now we've swapped the order of logging, event recording, and status updating for the non-successful paths. For the non-successful paths, the logging and event recording should happen **before** updating the status, so that the log and event is created regardless of whether the status update was successful. The logging and event recording should happen after updating the status **only** for the successful path so that we only inform of success when it is actually entirely successful.

Reply:
It seems a bit strange to log and record an error that wasn't actually represented in the resource state. For instance, we could potentially get into the following situation:
- we get an error from the broker
- the resource has hit its reconciliation retry duration and is being marked as failed
- the resource update command fails
- we retry reconciliation
- we get a success response from the broker
- the resource update succeeds

Now, we'd have in our logs and event records that the resource hit the reconciliation duration timeout, but our resource would show it completed everything fine. I think I'd rather the reconciler just throw up its hands and be like "forget the last thing that happened; I'm just going to try again" so that everything is consistent.

Reply:
My contention is that the error that the Controller is logging happened, regardless of whether the status was updated successfully or not. But, really, I don't care too much either way. I won't hold up this PR bickering about the order of logging.